src/OrchardCore/OrchardCore/Shell/ShellHost.cs

```diff
@@ -144,6 +144,7 @@ public async Task<ShellScope> GetScopeAsync(ShellSettings settings)
         public async Task UpdateShellSettingsAsync(ShellSettings settings)
         {
             settings.VersionId = IdGenerator.GenerateId();
+            settings.TenantId ??= IdGenerator.GenerateId();
```

Comment:
This is the only part that I'm going to need some time to think about. I think we need this setting to be generated once and not be able to change it afterward. This means that if I'm passing an arbitrary ShellSettings object without a TenantId taken from the actual .json file it will try and generate a new TenantId. This is tricky. I would somehow prefer that it be set inside the ISetupService only once and that this method never adds any chance to morph it on runtime.

Reply:
When we 1st create the Default tenant (setup context) we call this method. When we create a tenant from the admin UI, API or WF we call this method. When we edit a tenant we load the current settings (including the TenantId) to be updated. When we setup a tenant we load the current settings (uninitialized state but already including the TenantId when it was created) from which we create an internal settings (in the initializing state) but still including the TenantId.

Setting the TenantId as soon as we create a tenant allows to add setup features (enabled in the setup context) that may need the TenantId (e.g. Redis cache).

> This means that if I'm passing an arbitrary ShellSettings object without a TenantId taken from the actual .json file it will try and generate a new TenantId.

If the related tenant is being created we set its TenantId and then first save its settings.

For updating you should 1st load the settings of the related tenant (including the already set TenantId) then update it before passing it here, otherwise it would be inconsistent also for other properties.

Reply:
Forgot to say that also on AutoSetup, we first create a given tenant and save its settings by calling `UpdateShellSettingsAsync()` which will set its `TenantId` (before setup it afterwards)